pelican/utils.py

```diff
@@ -600,7 +635,9 @@ def truncate_html_words(s, num, end_text="…"):
     return out
 
 
-def process_translations(content_list, translation_id=None):
+def process_translations(
+    content_list: List["Content"], translation_id: Union[str, Set[str], None] = None
+) -> Tuple[List["Content"], List["Content"]]:
```

Comment:
Do these (`"Content"`) need to be strings? If I remember correctly, this is not needed anymore after python 3.7?

Reply:
> Do these (`"Content"`) need to be strings? If I remember correctly, this is not needed anymore after python 3.7?

As far as I know this is implemented with [PEP 563: Postponed Evaluation of Annotations](https://docs.python.org/3/whatsnew/3.7.html#pep-563-postponed-evaluation-of-annotations) and still requires `from __future__ import annotations`.

Reply:
Indeed it requires `from __future__ import annotations`. Otherwise a `NameError` will be raised. And because `pelican.contents` imports `pelican.utils` we can't remove the `if TYPE_CHECKING` gate here. Otherwise we create a circular import.

Reply:
I've accidentally pushed changes without the quotes. See https://github.com/getpelican/pelican/actions/runs/7623051907/job/20762296119 for the outcome

Reply:
Ah yes, removing quotes would require `__future__` import. Personally, I prefer `__future__` and non-quoted version, for simply maintainability reasons. If that becomes unnecessary later, it's easier to remove one line of import, instead of individually removing quotes from multiple places.

For `Union[..., None]` and `Optional[...]`, I know they are same, but you used `Optional[...]` everywhere else so I suggested only to continue that for consistency :). I gave up on making type hints readable. I think they make code equally unreadable (or readable depending on how you look), beyond anything simple. So I don't have a strong preference either way, other than keeping the same style throughout the project.

Reply:
I've changed it to `translation_id: Optional[Union[str, Collection[str]]]`. The following code uses `len` on `translation_id`. Therefore it must be a child of `Sized` and `Iterable` doesn't fit. Normally I use `Sequence` in this case but I noticed that `set` isn't a `Sequence` and therefore re-assigning a set on line 664 wouldn't work. Instead a `Collection` works.

Additionally I've re-read [PEP 563](https://peps.python.org/pep-0563/) and [PEP 649](https://peps.python.org/pep-0649/) and using `from __future__ import annotations` should be fine here as dataclasses or pydantic aren't in use here. With Python >= 3.13 the import can be safely removed again.